Adding new XML generation tool and example confs.

christinaholtNOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
Here I've added an xml_generator Python utility that will take in YAML files and write an XML. I have not yet documented the tool thoroughly, but wanted to get a bit of feedback from reviewers on the design ideas/concepts. 



## TESTS CONDUCTED: 
I have run the following command to "eyeball" the output format that it looks as I'd expect. 

`python xml_creator.py -c ./templates/rrfs_example.yml -o ./example_workflow.xml`

## DEPENDENCIES:
This will require Python v 3.9 to run. And I can follow up with a conda environment definition if needed to help you run this on your own.

## DOCUMENTATION:
TBD

christinaholtNOAA committed:
Adding new XML generation tool and example confs.

mark-a-potts committed:
Add Gaea as a supported platform for the regional_workflow (#734)

* Updates to port regional workflow to gaea

* Temp change with -v as batch option

* new fixes for gaea/slurm

* Updated time for make lbcs

* added TEST data directory path

* Update gaea.sh

* fixes for PR

gsketefian committed:
Add more parameters to CSV file containing WE2E test info (#740)

## DESCRIPTION OF CHANGES: 
The script/function `get_WE2Etest_names_subdirs_descs.sh` (which is called from `run_WE2E_tests.sh` if needed) creates a CSV (Comma-Separated Value) file named `WE2E_test_info.csv` that contains information about the WE2E tests.  Currently, this CSV file contains only 3 columns: the test name, the names of any alternate names for the test, and the test description.  In order to have a more complete summary of the WE2E tests, this PR modifies `get_WE2Etest_names_subdirs_descs.sh` so that additional information is included in the CSV file.  This additional information consists of the values of the following experiment variables for each test:
```
PREDEF_GRID_NAME
CCPP_PHYS_SUITE
EXTRN_MDL_NAME_ICS
EXTRN_MDL_NAME_LBCS
DATE_FIRST_CYCL
DATE_LAST_CYCL
CYCL_HRS
INCR_CYCL_FREQ
FCST_LEN_HRS
LBC_SPEC_INTVL_HRS
NUM_ENS_MEMBERS
```
In addition, the script uses this information to calculate the number of times each test calls the forecast model (e.g. if the test uses 3 different cycle dates, then the forecast model will be called 3 times; if it is an ensemble test for a single cycle, the test will call the forecast model as many times as the number of ensemble members).  

## TESTS CONDUCTED: 
The script `run_WE2E_tests.sh` was called that in turn calls `get_WE2Etest_names_subdirs_descs.sh`.  This created a new CSV file that contained the new fields (columns).  The CSV file was imported into Google Sheets (using "|" as the field/column separator) and looked correct.

## DOCUMENTATION:
The documentation is for the most part already within the `get_WE2Etest_names_subdirs_descs.sh`.  This PR slightly modifies that documentation to update it.

chan-hoo committed:
Update directory structure of NCO mode (#743)

* update vertical structure of NCO mode

* update sample script for nco

* Fix typo on write component of new RRFS CONUS

natalie-perlin committed:
Default CCPP physics option is FV3_GFS_v16 (#746)

* Updated the default CCPP physics option to FV3_GFS_v16

* Updated the default CCPP physics option to FV3_GFS_v16 in config_defaults.sh

Co-authored-by: Natalie Perlin <Natalie@Natalies-MacBook-Air.local>

danielabdi-noaa reviewed (commented):
I have tested it in a docker container (only place i can get python 3.9) and confirm it generates an xml file from the given sample yaml file.

danielabdi-noaa commented on ush/xml_creator.py:
```diff
@@ -0,0 +1,267 @@
+    return ' '.join([str(i) for i in arg])
+
+
+yaml.add_constructor('!cycstr', cycstr, Loader=yaml.SafeLoader)
+yaml.add_constructor('!include', include, Loader=yaml.SafeLoader)
+yaml.add_constructor('!startstopfreq', startstopfreq, Loader=yaml.SafeLoader)
```
I had one difficult to track bug with yaml constructors being "global". I belive you had a `!join` tag defined in `set_namelist.py` for joining paths and I had the same tag name defined for joining strings (needed that for `config_defaults.yaml) and this resulted in a conflict, adding "/" when my intention was to join strings. I wonder if there is a way to add yaml constructors to be local to avoid such kind of conflicts ...

christinaholtNOAA replied:
That's an interesting namespace issue. 

I'm not sure about the globals, but I am kind of curious if it would make sense to have a "unified" yaml parser that we could import/use in any of the tools that are ingesting a YAML. That may make it easier to deal with the conflicts.

danielabdi-noaa commented:
I noticed that generated xml file sometimes escapes ampersand when it is referencing an entity, for example, `&amp;SCHED` instead of directly referencing as `&SCHED`. I am not sure if they both do the same thing so just pointing out a minor difference ...

danielabdi-noaa commented on ush/templates/coldstart.yaml:
```diff
@@ -0,0 +1,156 @@
+  envars:
+    <<: *default_vars
+    SLASH_ENSMEM_SUBDIR: '/'
+  dependency:
+    and: &make_ics_deps
+      taskdep_get_extrn:
+        attrs:
+          task: get_extrn_ics
+      or_grid:
+        datadep_complete:
+          attrs:
+            age: 00:00:00:05
+          text: '&LOGDIR;/&MAKE_GRID_TN;_task_complete.txt'
+        streq:
+          left: '&RUN_TASK_MAKE_GRID;'
+          right: 'FALSE'
+      or_orog:
+        datadep_complete:
+          attrs:
+            age: 00:00:00:05
+          text: '&LOGDIR;/&MAKE_OROG_TN;_task_complete.txt'
+        streq:
+          left: '&RUN_TASK_MAKE_OROG;'
+          right: 'FALSE'
+      or_sfc_climo:
+        datadep_complete:
+          attrs:
+            age: 00:00:00:05
+          text: '&LOGDIR;/&MAKE_SFC_CLIMO_TN;_task_complete.txt'
+        streq:
+          left: '&RUN_TASK_MAKE_SFC_CLIMO;'
+          right: 'FALSE'
```
I think this is neat! If I understand correctly, every tag in the dependency dictionary is accounted for recursively, and is split with the first hyphen to get the corresponding rocoto tag, then form the xml statement using additional attributes.

christinaholtNOAA replied:
That's right.

danielabdi-noaa commented on ush/templates/coldstart.yaml:
```diff
@@ -0,0 +1,156 @@
+          attrs:
+            task: run_fcst
+        and:
+          datadep_dyn:
+            text: !cycstr {value: '&CYCLE_BASEDIR;/@Y@m@d@H/dynf#fhr#.nc'}
+            attrs:
+              age: '05:00'
+          datadep_phy:
+            text: !cycstr {value: '&CYCLE_BASEDIR;/@Y@m@d@H/phyf#fhr#.nc'}
+            attrs:
+              age: '05:00'
```
Is the second part of `datadep_dyn/phys/complete` etc parsed and used ? I am assuming reason why it is not just `datadep` is for the readablity of the yaml file?

christinaholtNOAA replied:
It is _not_ used. It's partially for readability, but also because you can't have repeated keys in the dict.

danielabdi-noaa committed:
Adds an alternative python workflow generation path (#698)

* Workflow in python starting to work.

* Use new python_utils package structure.

* Some bug fixes.

* Use uppercase TRUE/FALSE in var_dfns

* Use config.sh by default.

* Minor bug fixes.

* Remove config.yaml

* Update to the latest develop

* Remove quotes from numbers in predef grid.

* Minor bug fix.

* Move validity checker to the bottom of setup

* Add more unit tests.

* Update with python_utils changes.

* Update to latest develop additions (Need to re-run regression test)

* Use set_namelist and fill_jinja_template as python functions.

* Replace sed regex searches with python re.

* Use python realpath.

* Construct settings as dictionary before passing to fill_jinja and set_namelist

* Use yaml for setting predefined grid parameters.

* Use xml parser for ccpp phys suite definition file.

* Remove more run_command calls.

* Simplify some func argument processing.

* Move different config format parsers to same file.

* Use os.path.join for the sake of macosx

* Remove remaining func argument processing via os.environ.

* Minor bug fix in set_extrn_mdl_params.sh

* Add suite defn in test_data.

* Minor fixes on unittest on jet.

* Simplify boolean condition checks.

* Include old in renaming of old directories

* Fix conflicting yaml !join tag for paths and strings.

* Bug fix with setting sfcperst dict.

* Imitate "readlink -m" with os.path.realpath instead of os.readlink

* Don't use /tmp as that is shared by multiple users.

* Bug fix with cron line, maintain quotes around TRUE/FALSE.

* Update to latest develop (untested)

* Bug fix with existing cron line and quotes.

* Bug fix with case-sensitive MACHINE name, and empty EXPT_DIR.

* Update to latest develop

* More updates.

* Bug fix thanks to @willmayfield! Check both starting/ending
characters are brackets for shell variable to be considered an array.

* Make empty EXPT_BASEDIR workable.

* Update to latest develop

* Update in predef grid.

* Check f90nml as well.

Co-authored-by: Daniel Abdi <dabdi@Orion-login-2.HPC.MsState.Edu>

chan-hoo committed:
Fix typo and crontab issue on wcoss dell in workflow python scripts (#750)

* Fix typo and failure on wcoss

* fix new line issue on wcoss dell

* remove capture_output

* Get USER from environment

Co-authored-by: Daniel Abdi <daniel.abdi@noaa.gov>

EdwardSnyder-NOAA committed:
Add new WE2E configs (#748)

## DESCRIPTION OF CHANGES: 
Added two new WE2E config files for the Sub-CONUS Indianapolis domain to support the upcoming SRW release. 

In addition, modified the external data used in the `config.specify_EXTRN_MDL_SYSBASEDIR_ICS_LBCS.sh` to match more common datasets used in the WE2E testing process. 

## TESTS CONDUCTED: 
Successfully ran the new WE2E tests (`config.SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_HRRR.sh`, `config.SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta.sh`) and `config.specify_EXTRN_MDL_SYSBASEDIR_ICS_LBCS.sh` on NOAA Parallel Works AWS instance.

## DEPENDENCIES:
None.

## DOCUMENTATION:
No documentation changes are required.

ywangwof committed:
Added a fixed WoF grid and the python tool to determine the write component parameters (#733)

* Added a fixed WoF grid and the python tool to determine the write component parameters

* Update set_predef_grid_params.sh

* Renamed file as recommended and removed unused lines

* Modified comment

Co-authored-by: JeffBeck-NOAA <55201531+JeffBeck-NOAA@users.noreply.github.com>
Co-authored-by: WYH@MBP <yunheng.wang@4374-ywang.winstorm.nssl>

chan-hoo committed:
Replace env with modulefiles in scripts (#752)

* change env to mod

* update we2e script

christinaholtNOAA committed:
Updating some docs in the code.

gsketefian committed:
WE2E script improvements for usability (#745)

## DESCRIPTION OF CHANGES: 
* Modifications to `run_WE2E_tests.sh`:
  * Add examples to help/usage statement
* Modifications to `check_expts_status.sh`:
  * Add arguments list that can be processed by `process_args`
  * Add new optional arguments:  `num_log_lines`, `verbose`
  * Include a help/usage message

## TESTS CONDUCTED:
* Ran `run_WE2E_tests.sh --help` from the command line and got the expected help message.
* Ran `check_expts_status.sh --help` from the command line and got the expected help message.
* Used `run_WE2E_tests.sh` to run a set of 2 WE2E tests -- works as expected.
* Used `check_expts_status` to check on the status of the 2 tests run above and got the expected status message.
 
## DEPENDENCIES:
PR #[241](https://github.com/ufs-community/ufs-srweather-app/pull/241)

## DOCUMENTATION:
A lot of this PR is documentation in the scripts.  There is an accompanying documentation PR #[241](https://github.com/ufs-community/ufs-srweather-app/pull/241) into ufs-srweather-app.

mkavulich committed:
Standardize static data across Tier-1 platforms; fix and improve IC and LBC data retrieval (#744)

gsketefian committed:
Bug fixes (grid size + suppress screen output from module load) (#756)

## DESCRIPTION OF CHANGES: 
1) Adjust y-direction size of write-component grid of `SUBCONUS_Ind_3km` predefined grid from 195 to 197 (this was just an oversight in PR #725 ).
2) Redirect output of module load in launch script (`launch_FV3LAM_wflow.sh`) to `/dev/null` to avoid unwanted screen output (which was introduced in PR #[238](https://github.com/ufs-community/ufs-srweather-app/pull/238) in ufs-srweather-app and is about how to load the `regional_workflow` environment and is not relevant in this context).

## TESTS CONDUCTED: 
1) Plotted the `SUBCONUS_Ind_3km` grid to ensure it has correct size (it does).
2) Manually ran `launch_FV3LAM_wflow.sh` from the command line to verify that screen output is suppressed (it is).

JeffBeck-NOAA committed:
Update default SPP ISEED array in config_defaults.sh to use unique values (#759)

* Modify RRFS North America 3- and 13-km domain configuration and WE2E test.

* Modify default ISEED values for SPP

* Fix grid in WE2E test

chan-hoo committed:
Update workflow python scripts (#760)

* update python scripts

chan-hoo committed:
Change output file name of run_post to meet NCO standards (#758)

* change output file name

* change variable name

* update python script

* remove duplicates

* add a check for empty variables

* move variable to common area

* clean up unnecessary comments

* update scripts

* remove duplicate

* update python scripts

* fix user-staged dir path issue in python script

chan-hoo committed:
Add POST_OUTPUT_DOMAIN_NAME to WE2E tests for new grids (#763)

* Add new var to we2e tests for new grids

* rename we2e tests for custom grid

* remove unnecessary $

gsketefian committed:
Modifications to `CODEOWNERS` file (#757)

* Add @gspetro-NOAA, @natalie-perlin, and @EdwardSnyder-NOAA to CODEOWNERS so they are notified of all PRs and can review them.

* Remove duplicates in CODEOWNERS; remove users who will no longer be working with the repo.

christinaholtNOAA committed:
Adding a python utility for summarizing compute. (#769)

Adds a utility that summarizes Rocoto database computational usage information.

danielabdi-noaa committed:
Add github actions for python unittests. (#747)

* Add github actions for python unittests.

* Include all python script in ush

* Skip defining QUILTING params when it is set to False

* Update py_workflow

* Update unittest for set_extrn_mdl_params.

* Updates from develop.

Co-authored-by: Daniel Shawul <dshawul@yahoo.com>

chan-hoo committed:
Update sample script for NCO mode (#771)

* update config.nco.sh

* Add comment

mark-a-potts committed:
Feature/noaacloud (#767)

* updates for noaacloud

* working version

* fixes for noaacloud

* added extra modules for post

mark-a-potts committed:
removed cheyenne-specific crontab editing section (#773)

christinaholtNOAA committed:
Pin down hera miniconda3 module file version. (#770)

Pin down the version of miniconda3 on Hera, and do not append to the module path.

EdwardSnyder-NOAA committed:
update staged data dir (#774)

EdwardSnyder-NOAA committed:
update IC/LBC file paths to match new IC/LBC file structure (#766)

* update IC/LBC file paths to match new IC/LBC file structure

* change forecast length

* Update config.community.sh

These changes reflects the SRW team's preference for a 12-hours forecast. The gfs.t18z.pgrb2.0p25.f012 file will need to be added to all Level 1 systems.

Co-authored-by: Gillian Petro <96886803+gspetro-NOAA@users.noreply.github.com>

natalie-perlin committed:
add a machine file for MacOS (#777)

Co-authored-by: Natalie Perlin <perlin.natalie@gmail.com>

gsketefian committed:
Add columns for relative test and time step to CSV file containing WE2E test info (#776)

## DESCRIPTION OF CHANGES: 
This PR modifies the script `get_WE2Etest_names_subdirs_descs.sh` that creates the CSV file containing test information the columns for relative cost and time step (`DT_ATMOS`).  This allows users to have a better idea of what the cost of each test is (the relative cost is defined such that 1 unit represents the cost of running a single 6-hour forecast on the `RRFS_CONUS_25km` grid).  

Other related changes:
* Add the flag `generate_csv_file` to the `run_WE2E_tests.sh` script to allow users to skip the CSV file generation step (since it can take a while).
* Use arguments for the function `set_predef_grid_params()` so that it can be called not just from the experiment generation scripts but also others (in this case from `get_WE2Etest_names_subdirs_descs.sh`).  Necessary for this PR.
* Move some constants (`NH0`, `NH3`, `NH4`) from `config_defaults.sh` to `constants.sh` to make it easier to access them.
* Place the contents of `constants.sh` at the start of each experiment's `var_defns.sh` so that these constants are available to the j-jobs and ex-scripts when `var_defns.sh` is sourced.
* Use variable `valid_vals_BOOLEAN` in many more places (which allows many other `valid_vals_...` variables in `valid_param_vals.sh` and `setup.sh` to be removed).
* Several other bug fixes.

## TESTS CONDUCTED: 
The following tests (with results) were conducted on Hera.  There were only three (previously existing) failures.
```
* grid_CONUS_25km_GFDLgrid_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16:               SUCCESS
* grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp:           SUCCESS
* grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional:  SUCCESS
* grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2:                 SUCCESS
* grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16:                   SUCCESS
* grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR:                         SUCCESS
* grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_2017_gfdlmp:           SUCCESS
* grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2:                 SUCCESS
* grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v16:                   FAILURE
* grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_HRRR:                            SUCCESS
* grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta:                     SUCCESS
* grid_RRFS_CONUScompact_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16:            SUCCESS
* grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR:                   SUCCESS
* grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_RRFS_v1beta:            SUCCESS
* grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR:                    SUCCESS
* grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1alpha:            SUCCESS
* grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta:             SUCCESS
* grid_SUBCONUS_Ind_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16:                  SUCCESS
* grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_HRRR:                          SUCCESS
* grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta:                   SUCCESS
* GST_release_public_v1:                                                       SUCCESS
* MET_ensemble_verification:                                                   SUCCESS
* MET_verification:                                                            SUCCESS
* community_ensemble_008mems:                                                  SUCCESS
* community_ensemble_2mems:                                                    SUCCESS
* custom_ESGgrid:                                                              SUCCESS
* custom_GFDLgrid:                                                             SUCCESS
* custom_GFDLgrid__GFDLgrid_USE_GFDLgrid_RES_IN_FILENAMES_eq_FALSE:            SUCCESS
* custom_GFDLgrid__GFDLgrid_USE_GFDLgrid_RES_IN_FILENAMES_eq_TRUE:             SUCCESS
* deactivate_tasks:                                                            SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2019061200:                   SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2019101818:                   SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2020022518:                   SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2020022600:                   SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_grib2_2021010100:                   SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio:                             SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200:                  SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019101818:                  SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2020022518:                  SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2020022600:                  SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2021010100:                  SUCCESS
* get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_netcdf_2021062000:                  SUCCESS
* get_from_HPSS_ics_GSMGFS_lbcs_GSMGFS:                                        SUCCESS
* get_from_HPSS_ics_HRRR_lbcs_RAP:                                             SUCCESS
* get_from_HPSS_ics_RAP_lbcs_RAP:                                              SUCCESS
* get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio:                           SUCCESS
* inline_post:                                                                 SUCCESS
* nco_ensemble:                                                                SUCCESS
* nco_inline_post:                                                             SUCCESS
* pregen_grid_orog_sfc_climo:                                                  SUCCESS
* specify_DOT_OR_USCORE:                                                       SUCCESS
* specify_DT_ATMOS_LAYOUT_XY_BLOCKSIZE:                                        SUCCESS
* specify_EXTRN_MDL_SYSBASEDIR_ICS_LBCS:                                       SUCCESS
* specify_RESTART_INTERVAL:                                                    SUCCESS
* specify_template_filenames:                                                  SUCCESS
* subhourly_post:                                                              FAILURE
* subhourly_post_ensemble_2mems:                                               FAILURE
```

## DOCUMENTATION:
Documentation is updated via PR #[278](https://github.com/ufs-community/ufs-srweather-app/pull/278) in the ``ufs-srweather-app`` repo.

gsketefian committed:
Bug fix for calculation of number of time steps for a given WE2E test. (#782)

## DESCRIPTION OF CHANGES: 
This PR fixes a bug in the way the number of time steps (`num_time_steps` in `get_WE2Etest_names_subdirs_descs.sh`) is calculated for each WE2E test when creating the CSV file containing information about the available WE2E tests.

## TESTS CONDUCTED: 
Reran the `run_WE2E_tests.sh` script (which in turn calls `get_WE2Etest_names_subdirs_descs.sh`), and a CSV file containing the adjusted numbers was successfully created.

## DOCUMENTATION:
None needed.

gsketefian committed:
Bugfix:  Pass domain name to python plotting scripts (#783)

## DESCRIPTION OF CHANGES: 
PR #[763](https://github.com/ufs-community/regional_workflow/pull/763) introduced the experiment variable `POST_OUTPUT_DOMAIN_NAME` and used it to modify the names of the UPP output files.  This PR adds `POST_OUTPUT_DOMAIN_NAME` as an argument to the python plotting scripts to enable those scripts to form the proper UPP output file names to read in.  It also modifies the batch scripts (`qsub_job.sh`, `qsub_job_diff.sh`, `sq_job.sh`, and `sq_job_diff.sh`) to add the new argument to the calls to the python plotting scripts that are called.

## TESTS CONDUCTED: 
Both python plotting scripts `plot_allvars.py` and `plot_allvars_diff.py` were run on Hera on preexisting UPP output files and completed successfully.  The batch scripts `sq_job.sh` and `sq_job_diff.sh` were also run successfully on Hera (`qsub_job.sh` and `qsub_job_diff.sh` were not run on e.g. Cheyenne).

## DEPENDENCIES:
PR # in `ufs-srweather-app`.

## DOCUMENTATION:
Modifications to the documentation are in PR #.

## ISSUE: 
Partially resolves issue #781 (still need to introduce tests in `regional_workflow` that run the plotting scripts).

## CONTRIBUTORS: 
@mark-a-potts pointed out the problem and first made the necessary fixes in the `NOAA-EPIC` fork of `regional_workflow`; @gsketefian introduced those changes to this PR and made edits to the comments and documentation.

chan-hoo committed:
Update scripts to work with the latest hashes of UFS_UTILS and UPP (#775)

* update input namelist of chgres_cube

* update diag_table templates

* update scripts

* back to original

* specify miniconda version on Jet

EdwardSnyder-NOAA committed:
adding new e2e tests (#792)

chan-hoo committed:
[develop] Update scripts to run with the latest hash of ufs weather model (#796)

* update new variables introduced in uwm

* update python script

* update template for HRRR suite

willmayfield committed:
[develop] Update MET/METplus versions and configuration files (#779)

* remove unused module loads for MET

* update met version in machine files

* changes to pointstat/pb2nc bufr mapping

* add NA to retain bufr variables without qc markers

* updated bufr variable list

* change bufr var list

* Change versions to MET 10.1.1 and METplus 4.1.1

gspetro-NOAA committed:
add a linux.sh machine file (a template) (#798) (#803)

Co-authored-by: Natalie Perlin <68030316+natalie-perlin@users.noreply.github.com>

gsketefian committed:
[develop] Rename variables related to GFDLgrid-type grids (#787)

## DESCRIPTION OF CHANGES: 
For clarity, rename variables related to GFDLgrid-type grids:
1) Rename `GFDLgrid_RES` to `GFDLgrid_NUM_CELLS`.
2) Rename `GFDLgrid_USE_GFDLgrid_RES_IN_FILENAMES` to `GFDLgrid_USE_NUM_CELLS_IN_FILENAMES`.
3) Rename WE2E tests on GFDLgrid-type custom grids according to variable renaming above.

Related improvements:
* In `link_fix.sh`, remove creation of unnecessary symlink when using a GFDLgrid-type grid with `GFDLgrid_USE_NUM_CELLS_IN_FILENAMES` is set to `"FALSE"`.
* In WE2E test configuration files that use custom grids, move the definition of the variable `POST_OUTPUT_DOMAIN_NAME` to the beginning of the custom grid definition because it is really the name of the custom native grid, and it's most appropriate to place that at the beginning of the grid definition section.
* Reduce `WTIME_MAKE_OROG` from 1 hour to 20 minutes since on Hera, 1 hour causes the task to wait quite a while in the queue.  This used to be 20 minutes but was changed to 1 hour because 20 minutes was not sufficient to complete the task on `WCOSS_DELL_P3`.  Thus, reset it to 1 hour only in the machine file for `WCOSS_DELL_P3`.
* Minor fixes to comments in python plotting scripts.
* Edits to other comments.

## TESTS CONDUCTED: 
Ran the four WE2E tests that use custom grids (listed below) on Hera.  All were successful.
* `custom_ESGgrid`
* `custom_GFDLgrid`
* `custom_GFDLgrid__GFDLgrid_USE_NUM_CELLS_IN_FILENAMES_eq_FALSE` (previously `custom_GFDLgrid__GFDLgrid_USE_GFDLgrid_RES_IN_FILENAMES_eq_FALSE`)
* `custom_GFDLgrid__GFDLgrid_USE_NUM_CELLS_IN_FILENAMES_eq_TRUE` (previously `custom_GFDLgrid__GFDLgrid_USE_GFDLgrid_RES_IN_FILENAMES_eq_TRUE`)

## DEPENDENCIES:
PR #[310](https://github.com/ufs-community/ufs-srweather-app/pull/310) in `ufs-srweather-app`.

## DOCUMENTATION:
Updates to documentation are in PR #[310](https://github.com/ufs-community/ufs-srweather-app/pull/310) in `ufs-srweather-app`.

christinaholtNOAA committed:
Remove un-readable path to systemdir on Hera. (#806)

This is a bug fix for running the WE2E tests on Hera. The non-existence of this path is causing failures after the path has become unreadable.

EdwardSnyder-NOAA committed:
[develop] Add MET/METplus to Jet (#799)

* add metplus paths

* add run_vx.local for jet

Co-authored-by: Edward Snyder <Edward.Snyder@noaa.com>

christinaholtNOAA committed:
Merge remote-tracking branch 'origin/develop' into xml_generator

christinaholtNOAA committed:
Adding configs for cold start and prep.

christinaholtNOAA committed:
Better comments and CLI parsing.

christinaholtNOAA committed:
Adding post workflow.

christinaholtNOAA committed:
Clean up existing workflows.

christinaholtNOAA committed:
Adding draft for verification workflows.

christinaholtNOAA committed:
Cleanup jobname, log, and default_task naming.
